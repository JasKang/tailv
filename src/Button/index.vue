<script setup lang="ts">
import { computed, type PropType } from 'vue'

import Loading from '../Icon/Loading.vue'
import { cvar, type ColorAlias } from '@/utils/theme'

defineOptions({ name: 'Button' })
const emit = defineEmits<{ click: [Event] }>()
const slots = defineSlots<{ default?(_: {}): any; icon?(_: {}): any }>()
const props = defineProps({
  variant: {
    type: String as PropType<'solid' | 'soft' | 'outline' | 'text' | 'link' | 'pure'>,
    default: 'outline',
  },
  size: {
    type: String as PropType<'sm' | 'md' | 'lg'>,
    default: 'md',
  },
  color: {
    type: String as PropType<'primary' | 'success' | 'warning' | 'danger' | 'none'>,
    default: 'none',
  },
  pill: Boolean,
  square: Boolean,
  block: Boolean,
  loading: Boolean,
  disabled: Boolean,
})
</script>
<template>
  <button
    type="button"
    class="inline-flex appearance-none items-center justify-center whitespace-nowrap text-center font-medium outline-none transition-all focus:z-10 focus:outline-none disabled:pointer-events-none disabled:opacity-50"
    :class="[
      {
        outline: `border-2 shadow-sm focus:ring-2 focus:ring-offset-2 ${
          {
            none: 'hover:bg-default-soft border-[--t-border] text-default focus:ring-primary',
            primary: 'hover:bg-primary-soft border-primary text-primary focus:ring-primary',
            success: 'hover:bg-success-soft border-success text-success focus:ring-success',
            warning: 'hover:bg-warning-soft border-warning text-warning focus:ring-warning',
            danger: 'hover:bg-danger-soft border-danger text-danger focus:ring-danger',
          }[props.color]
        }`,
        solid: `shadow-sm focus:ring-2 focus:ring-offset-2 ${
          {
            none: 'hover:bg-default-soft border border-[--t-border] focus:ring-primary',
            primary: 'text-primary-fg hover:bg-primary-hover bg-primary focus:ring-primary',
            success: 'text-success-fg hover:bg-success-hover bg-success focus:ring-success',
            warning: 'text-warning-fg hover:bg-warning-hover bg-warning focus:ring-warning',
            danger: 'text-danger-fg hover:bg-danger-hover bg-danger focus:ring-danger',
          }[props.color]
        }`,
        soft: `focus:ring-2 focus:ring-offset-2 ${
          {
            none: 'bg-default-soft hover:bg-default-soft-hover text-default focus:ring-primary',
            primary: 'bg-primary-soft hover:bg-primary-soft-hover text-primary focus:ring-primary',
            success: 'bg-success-soft hover:bg-success-soft-hover text-success focus:ring-success',
            warning: 'bg-warning-soft hover:bg-warning-soft-hover text-warning focus:ring-warning',
            danger: 'bg-danger-soft hover:bg-danger-soft-hover text-danger focus:ring-danger',
          }[props.color]
        }`,
        text: `focus:ring-2 focus:ring-offset-2 ${
          {
            none: 'hover:bg-default-soft focus:ring-default-soft-hover text-default',
            primary: 'hover:bg-primary-soft focus:ring-primary-soft-hover text-primary',
            success: 'hover:bg-success-soft focus:ring-success-soft-hover text-success',
            warning: 'hover:bg-warning-soft focus:ring-warning-soft-hover text-warning',
            danger: 'hover:bg-danger-soft focus:ring-danger-soft-hover text-danger',
          }[props.color]
        }`,
        link: `underline-offset-4 hover:underline ${
          {
            none: 'text-default',
            primary: 'text-primary',
            success: 'text-success',
            warning: 'text-warning',
            danger: 'text-danger',
          }[props.color]
        }`,
        pure: '',
      }[props.variant],
      {
        sm: `h-7 text-xs ${props.square ? 'w-7' : 'px-2 py-1'}`,
        md: `h-9 text-sm ${props.square ? 'w-9' : 'px-3 py-1.5'}`,
        lg: `h-11 text-base ${props.square ? 'w-11' : 'px-4 py-2'}`,
      }[props.size],
      props.pill ? 'rounded-full' : 'rounded-md',
      !props.square && props.block ? 'w-full' : '',
    ]"
    :disabled
    @click="emit('click', $event)"
  >
    <template v-if="loading">
      <Loading class="h-[1em] w-[1em] animate-spin" :class="$slots.default && !square ? 'mr-1.5' : ''" />
    </template>

    <slot v-if="!square || !loading"></slot>
  </button>
</template>
