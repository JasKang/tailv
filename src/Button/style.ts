import { tw } from '../utils/tw'

// 'primary' | 'success' | 'warning' | 'danger'

function createVariantColor(variant: 'default' | 'ghost' | 'secondary', color?: string) {
  if (variant === 'ghost') {
    if (color) {
      return {
        '--btn-bg-color': '',
        '--btn-text-color': '',
        '--btn-border-color': '',
        '--btn-bg-color-hover': '',
        '--btn-text-color-hover': '',
        '--btn-border-color-hover': '',
      }
    } else {
      return `bg-transparent hover:bg-slate-100`
    }
  } else if ('secondary') {
    if (color) {
      return `bg-transparent text-${color} hover:bg-${color}-100`
    } else {
      return `bg-transparent hover:bg-slate-100`
    }
  } else {
    if (color) {
      return `bg-transparent text-${color} hover:bg-${color}-100`
    } else {
      return `border-slate-300 bg-white text-current enabled:hover:bg-slate-50`
    }
  }
}

export const createButtonCss = tw(
  `inline-flex appearance-none items-center justify-center border border-[--btn-border-color] bg-[--btn-bg-color] 
  text-sm font-medium text-[--btn-text-color] outline-[--btn-outline-color]
  transition-colors 
  focus-visible:outline
  focus-visible:outline-2 
  focus-visible:outline-offset-2 
  enabled:hover:border-[--btn-border-color-hover] 
  enabled:hover:bg-[--btn-bg-color-hover] 
  enabled:hover:text-[--btn-text-color-hover]
  `,
  {
    compoundVariants: [
      { size: 'sm', square: true, class: 'w-8' },
      { size: 'md', square: true, class: 'w-9' },
      { size: 'lg', square: true, class: 'w-10' },
      {
        variant: 'default',
        color: 'normal',
        class: `border-slate-300 bg-white text-current enabled:hover:bg-slate-50`,
      },
      {
        variant: 'default',
        color: 'primary',
        class: `border-transparent bg-primary-500 text-white enabled:hover:bg-primary-600`,
      },
      {
        variant: 'default',
        color: 'success',
        class: `border-transparent bg-success-500 text-white  enabled:hover:bg-success-600`,
      },
      {
        variant: 'default',
        color: 'warning',
        class: `border-transparent bg-warning-500 text-white  enabled:hover:bg-warning-600`,
      },
      {
        variant: 'default',
        color: 'danger',
        class: `border-transparent bg-danger-500 text-white  enabled:hover:bg-danger-600`,
      },

      {
        variant: 'secondary',
        color: 'normal',
        class: `border-transparent bg-slate-100 text-current enabled:hover:bg-slate-200`,
      },
      {
        variant: 'secondary',
        color: 'primary',
        class: `border-transparent bg-primary-100 text-primary-600 enabled:hover:bg-primary-200`,
      },
      {
        variant: 'secondary',
        color: 'success',
        class: `border-transparent bg-success-100 text-success-600  enabled:hover:bg-success-200`,
      },
      {
        variant: 'secondary',
        color: 'warning',
        class: `border-transparent bg-warning-100 text-warning-600  enabled:hover:bg-warning-200`,
      },
      {
        variant: 'secondary',
        color: 'danger',
        class: `border-transparent bg-danger-100 text-danger-600  enabled:hover:bg-danger-200`,
      },

      {
        variant: 'ghost',
        color: 'normal',
        class: `border-transparent bg-transparent text-current enabled:hover:bg-slate-200`,
      },
      {
        variant: 'ghost',
        color: 'primary',
        class: `border-transparent bg-transparent text-primary-600 enabled:hover:bg-primary-200`,
      },
      {
        variant: 'ghost',
        color: 'success',
        class: `border-transparent bg-transparent text-success-600  enabled:hover:bg-success-200`,
      },
      {
        variant: 'ghost',
        color: 'warning',
        class: `border-transparent bg-transparent text-warning-600  enabled:hover:bg-warning-200`,
      },
      {
        variant: 'ghost',
        color: 'danger',
        class: `border-transparent bg-transparent text-danger-600  enabled:hover:bg-danger-200`,
      },
    ],
    variants: {
      block: {
        false: '',
        true: 'w-full',
      },
      disabled: {
        false: 'cursor-pointer',
        true: 'cursor-not-allowed opacity-60',
      },
      pill: {
        false: 'rounded-md',
        true: 'rounded-full',
      },
      size: {
        lg: ['h-10 px-3.5 py-2.5', ''],
        md: ['h-9 px-3 py-2', ''],
        sm: ['h-8 px-2.5 py-1.5', ''],
      },
      square: {
        false: '',
        true: ['overflow-hidden px-0', 'px-2.5 px-3 px-3.5'],
      },

      // {
      //   variant: 'default',
      //   color: 'normal',
      //   class: `border-slate-300 bg-white text-current enabled:hover:bg-slate-50`,
      // },
      // {
      //   variant: 'default',
      //   color: 'primary',
      //   class: `border-transparent bg-primary-500 text-white enabled:hover:bg-primary-600`,
      // },

      // bg-[--btn-bg-color]
      // text-[--btn-text-color]
      // border-[--btn-border-color]
      // enabled:hover:bg-[--btn-bg-color-hover]
      // enabled:hover:text-[--btn-text-color-hover]
      // enabled:hover:border-[--btn-border-color-hover]

      variant: {
        default: 'shadow-sm',
        secondary: 'shadow-sm',
        ghost: '',
      },
      color: {
        normal: '',
        primary: '',
        success: '',
        warning: '',
        danger: '',
      },
    },
  }
)
