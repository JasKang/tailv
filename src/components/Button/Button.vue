<script setup lang="ts">
import { ref, computed, type PropType } from 'vue'
import { getBtnVars, useCls } from './styles'
import { useTheme } from '@/core/theme'
import LoadingIcon from '../svgs/LoadingIcon.vue'
import type { ColorKey } from '@/core/colors'

defineOptions({ name: 'TButton' })
const emit = defineEmits<{ click: [MouseEvent] }>()
const slots = defineSlots<{ default?(_: {}): any; icon?(_: {}): any }>()
const props = defineProps({
  variant: {
    type: String as PropType<'solid' | 'soft' | 'outline' | 'link' | 'subtle'>,
    default: 'solid',
  },
  color: {
    type: String as PropType<'primary' | 'success' | 'warning' | 'error' | ColorKey>,
  },
  size: {
    type: String as PropType<'xs' | 'sm' | 'md' | 'lg' | 'xl'>,
    default: 'md',
  },
  ring: {
    type: Boolean,
    default: true,
  },
  rounded: Boolean,
  square: Boolean,
  circle: Boolean,
  block: Boolean,
  loading: Boolean,
  disabled: Boolean,
})

const { getColorKey } = useTheme()

const cls = useCls(() => {
  return {
    variant: props.variant,
    size: props.size,
    ring: props.ring,
    rounded: props.rounded || props.circle,
    square: props.square || props.circle,
    block: props.block,
    disabled: props.disabled,
  }
})
const cssVars = computed(() => {
  return getBtnVars(props.variant, getColorKey(props.color || 'gray'))
})

const hasIcon = computed(() => !!slots.icon || props.loading)
const onClick = (e: MouseEvent) => {
  if (!props.disabled) {
    emit('click', e)
  }
}
</script>
<template>
  <button :class="cls" :style="cssVars" type="button" :disabled="disabled" @click="onClick">
    <i v-if="hasIcon" class="t-btn-icon h-[1em] w-[1em] scale-125 [&+*]:ml-[0.5em] [&>svg]:!h-full [&>svg]:!w-full">
      <LoadingIcon v-if="loading" class="animate-spin" />
      <slot v-else name="icon" />
    </i>
    <span v-if="slots.default" class="t-btn_body">
      <slot />
    </span>
  </button>
</template>
