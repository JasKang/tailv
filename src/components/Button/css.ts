import { type ArrayCSSInterpolation, css, type CSSInterpolation } from '@emotion/css'

export const cssName = () => {
  const cls = css([
    `
    --z-btn-text-color: inherit;
    --z-btn-bg-color: #fff;
    --z-btn-border-color: #fff;
    --z-btn-bg-color-hover: var(--z-btn-bg-color);
    --z-btn-border-color-hover: var(--z-btn-border-color);

    --z-btn-outline-color: var(--z-primary-color);
    --z-btn-h: calc(2.25rem + 2px);

    appearance: none;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    border-radius: 0.375rem;
    height: var(--z-btn-h);
    color: var(--z-btn-text-color);
    background-color: var(--z-btn-bg-color);
    text-align: center;
    font-weight: 500;
    border: none;
    outline: none;
    outline-offset: 2px;
    transition-property: color, background-color, border-color, text-decoration-color, fill, stroke;
    transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
    transition-duration: 150ms;

    .z-btn_icon {
      display: block;
      width: 1em;
      height: 1em;
      line-height: 1em;
      transform: scale3d(1.25, 1.25, 1);
      > svg {
        display: block;
        width: 100%;
        height: 100%;
      }
      + .z-btn_body{
        margin-left: 0.4em;
      }
    }
    .z-btn_body {
      display: flex;
      align-items: center;
      justify-content: center;
    }

    &:enabled {
      cursor: pointer;
      &:hover {
        background-color: var(--z-btn-bg-color-hover);
        border-color: var(--z-btn-border-color-hover);
      }
      &:focus-visible {
        outline: var(--z-btn-outline-color) solid 2px;
      }
    }

    &:disabled {
      cursor: not-allowed;
      opacity: 0.5;
    }
  `,
    ...genVariantStyle(),
    ...genSizeStyle(),
    ...genOtherStyle(),
  ])
  return { css: cls }
}

function genVariantStyle(): ArrayCSSInterpolation {
  return [
    {
      '&.z-btn-default': {
        'box-shadow': 'var(--shadow-sm)',
        border: '1px solid var(--z-btn-border-color)',
      },
      '&.z-btn-outline': {
        border: '1px solid var(--z-btn-border-color)',
      },
      '&.z-btn-solid': {
        'box-shadow': 'var(--shadow-sm)',
        color: '#fff',
      },
      '&.z-btn-soft': {
        'box-shadow': 'var(--shadow-sm)',
      },
      '&.z-btn-plain': {},
      '&.z-btn-link': {
        '&:enabled:hover': {
          'text-decoration': 'underline solid green 2px',
          'text-underline-offset': '4px',
        },
      },
    },
  ]
}

function genSizeStyle() {
  return [
    {
      '&.z-btn--xs': {
        '--z-btn-h': 'calc(1.75rem + 2px)',
        'font-size': '0.75rem',
        padding: '0 0.5rem',
      },
      '&.z-btn--sm': {
        '--z-btn-h': 'calc(2rem + 2px)',
        'font-size': '0.875rem',
        padding: '0 0.75rem',
      },
      '&.z-btn--md': {
        '--z-btn-h': 'calc(2.25rem + 2px)',
        'font-size': '1rem',
        padding: '0 1rem',
      },
      '&.z-btn--lg': {
        '--z-btn-h': 'calc(2.5rem + 2px)',
        'font-size': '1.125rem',
        padding: '0 1.25rem',
      },
      '&.z-btn--xl': {
        '--z-btn-h': 'calc(2.75rem + 2px)',
        'font-size': '1.25rem',
        padding: '0 1.5rem',
      },
    },
  ]
}

function genOtherStyle() {
  return [
    {
      '&.z-btn--block': {
        width: '100%',
      },
      '&.z-btn--pill': {
        'border-radius': '9999px',
      },
      '&.z-btn--square': {
        width: 'var(--z-btn-h)',
        padding: '0 !important',
      },
    },
  ]
}
