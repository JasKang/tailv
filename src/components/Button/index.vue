<script setup lang="ts">
import { ref, computed } from 'vue'
import { useCls } from './styles'

defineOptions({ name: 'TButton' })
const props = defineProps({})
const emit = defineEmits<{ click: [number] }>()
const slots = defineSlots<{ default(): any }>()

const cls = useCls(() => {
  return {
    variant: props.variant,
    color: props.color,
    size: props.size,
    rounded: props.rounded || props.circle,
    square: props.square || props.circle,
    block: props.block,
    disabled: props.disabled,
  }
})
</script>
<template>
  <div :class="cls"></div>
</template>
<style lang="scss"></style>
