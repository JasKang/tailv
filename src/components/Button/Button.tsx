import {
  computed,
  defineComponent,
  type ExtractPropTypes,
  type ExtractPublicPropTypes,
  type PropType,
  type SlotsType,
  type VNode,
} from 'vue'
import { useStyle } from './style'
import { useTheme } from '@/theme'

const props = {
  variant: {
    type: String as PropType<'solid' | 'soft' | 'outline' | 'link' | 'subtle'>,
    default: 'solid',
  },
  color: {
    type: String as PropType<'primary' | 'success' | 'warning' | 'error' | ColorKey>,
  },
  size: {
    type: String as PropType<'xs' | 'sm' | 'md' | 'lg' | 'xl'>,
    default: 'md',
  },
  ring: {
    type: Boolean,
    default: true,
  },
  rounded: Boolean,
  square: Boolean,
  circle: Boolean,
  block: Boolean,
  loading: Boolean,
  disabled: Boolean,
}

export type ButtonProps = ExtractPropTypes<typeof props>

export type ButtonPublicProps = ExtractPublicPropTypes<typeof props>

export type ButtonCssVars = {
  '--t-btn-text-color': string
  '--t-btn-text-color-hover': string
  '--t-btn-border-color': string
  '--t-btn-border-color-hover': string
  '--t-btn-bg': string
  '--t-btn-bg-hover': string
  '--t-btn-ring-color': string
}

export const Button = defineComponent({
  name: 'TButton',
  props,
  emits: {
    click: (payload: MouseEvent) => {},
  },
  slots: Object as SlotsType<{
    default: () => VNode
    item: { data: number }
  }>,
  setup(props, { slots, emit }) {
    const { getColorKey } = useTheme()

    const { cssVars, cls } = useStyle(() => {
      return {
        ...props,
        rounded: props.rounded || props.circle,
        square: props.square || props.circle,
      }
    })

    const hasIcon = computed(() => !!slots.icon || props.loading)
    const onClick = (e: MouseEvent) => {
      if (!props.disabled) {
        emit('click', e)
      }
    }
  },
})
