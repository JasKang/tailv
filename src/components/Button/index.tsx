import {
  computed,
  defineComponent,
  type ExtractPropTypes,
  type ExtractPublicPropTypes,
  type ObjectEmitsOptions,
  type PropType,
  type SlotsType,
  type VNode,
} from 'vue'

import type { UserColor } from '@/theme/colors'

import { LoadingIcon } from '../Icon'
import { useButtonStyle } from './style'

const props = {
  variant: {
    type: String as PropType<'default' | 'solid' | 'soft' | 'outline' | 'plain' | 'link'>,
    default: 'default',
  },
  color: {
    type: String as PropType<UserColor>,
    default: 'primary',
  },
  size: {
    type: String as PropType<'xs' | 'sm' | 'md' | 'lg' | 'xl'>,
    default: 'md',
  },
  pill: Boolean,
  square: Boolean,
  circle: Boolean,
  block: Boolean,
  loading: Boolean,
  disabled: Boolean,
}

export type ButtonInnerProps = ExtractPropTypes<typeof props>

export type ButtonProps = ExtractPublicPropTypes<typeof props>

export const Button = defineComponent({
  name: 'TButton',
  props,
  emits: {
    click: (payload: MouseEvent) => true,
  },
  slots: Object as SlotsType<{
    default: () => VNode
    icon: () => VNode
  }>,
  setup(props, { slots, emit, attrs }) {
    const ownStyle = useButtonStyle(() => props)

    const hasIcon = computed(() => !!slots.icon || props.loading)
    const onClick = (e: MouseEvent) => {
      if (!props.disabled) {
        emit('click', e)
      }
    }
    return () => (
      <button
        style={ownStyle.value.style}
        class={ownStyle.value.class}
        type="button"
        disabled={props.disabled}
        onClick={onClick}
      >
        {hasIcon.value && (
          <i class="t-btn-icon h-[1em] w-[1em] scale-125 [&+*]:ml-[0.5em] [&>svg]:!h-full [&>svg]:!w-full">
            {props.loading ? <LoadingIcon class="animate-spin" /> : slots.icon?.()}
          </i>
        )}
        {slots.default && <span class="t-btn_body flex items-center">{slots.default()}</span>}
      </button>
    )
  },
})
