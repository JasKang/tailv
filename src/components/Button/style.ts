import { computed, type MaybeRefOrGetter, toValue } from 'vue'

import { useTheme } from '@/theme'
import type { PaletteColor, VarColor } from '@/theme/colors'
import { clsVariants } from '@/utils/clst'
import { createStyleVar } from '@/utils/style'

import type { ButtonInnerProps } from '.'

type ButtonCssVars = {
  text: VarColor
  bg: VarColor
  border: VarColor
  textHover: VarColor
  bgHover: VarColor
  borderHover: VarColor
  ring: VarColor
}

const cssVars = createStyleVar<'t-btn', ButtonCssVars>('t-btn')

const createClass = clsVariants(
  `t-button inline-flex rounded-md shadow-sm text-center justify-center items-center font-medium transition-all 
  focus:outline-none h-[--t-btn-h]
  text-[--t-btn-text] bg-[--t-btn-bg] border-[--t-btn-border]
  hover:text-[--t-btn-textHover] hover:bg-[--t-btn-bgHover] hover:border-[--t-btn-borderHover]
  `,
  {
    variants: {
      variant: {
        default: ['border', ''],
        outline: ['border-[1.5px]', ''],
        solid: 'border-2',
        soft: 'border-2',
        plain: ['border-2', 'shadow-sm'],
        link: ['border-2 hover:underline hover:decoration-2 hover:underline-offset-2', 'shadow-sm'],
      },
      color: {
        primary: '',
        success: '',
        warning: '',
        error: '',
      },
      size: {
        xs: '[--t-btn-h:calc(1.75rem+2px)] text-xs/3 px-2',
        sm: '[--t-btn-h:calc(2rem+2px)] text-xs/4 px-3',
        md: '[--t-btn-h:calc(2.25rem+2px)] text-sm/5 px-4',
        lg: '[--t-btn-h:calc(2.5rem+2px)] text-base/6 px-5',
        xl: '[--t-btn-h:calc(2.75rem+2px)] text-base/7 px-6',
      },
      block: {
        true: 'w-full',
        false: '',
      },
      pill: {
        true: 'rounded-full',
        false: '',
      },
      square: {
        true: ['px-0 w-[--t-btn-h]', 'px-2 px-3 px-4 px-5 px-6'],
        false: '',
      },
      disabled: {
        true: 'cursor-not-allowed opacity-50 hover:-underline hover:text-[--t-btn-text] hover:bg-[--t-btn-bg] hover:border-[--t-btn-border]',
        false:
          'cursor-pointer focus:ring-2 focus:ring-[--t-btn-ring] focus:ring-offset-2 dark:focus:ring-offset-gray-900',
      },
    },
  }
)

function createStyle(variant: ButtonInnerProps['variant'], color: PaletteColor) {
  switch (variant) {
    case 'solid':
      return cssVars({
        text: `white`,
        bg: `${color}.500`,
        border: `transparent`,
        textHover: `white`,
        bgHover: `${color}.600`,
        borderHover: `transparent`,
        ring: `${color}.500`,
      })
    case 'soft':
      return cssVars({
        text: `${color}.500`,
        bg: `${color}.100`,
        border: `transparent`,
        textHover: `${color}.500`,
        bgHover: `${color}.200`,
        borderHover: `transparent`,
        ring: `${color}.500`,
      })
    case 'plain':
      return cssVars({
        text: `${color}.500`,
        bg: `transparent`,
        border: `transparent`,
        textHover: `${color}.500`,
        bgHover: `${color}.100`,
        borderHover: `transparent`,
        ring: `${color}.500`,
      })
    case 'link':
      return cssVars({
        text: `${color}.500`,
        bg: `transparent`,
        border: `transparent`,
        textHover: `${color}.500`,
        bgHover: `transparent`,
        borderHover: `transparent`,
        ring: `${color}.500`,
      })
    case 'outline':
      return cssVars({
        text: `${color}.500`,
        bg: 'white',
        border: `${color}.500`,
        textHover: 'white',
        bgHover: `${color}.500`,
        borderHover: `${color}.500`,
        ring: `${color}.500`,
      })
    default:
      return cssVars({
        text: `gray.700`,
        bg: 'white',
        border: `gray.300`,
        textHover: `gray.900`,
        bgHover: `gray.50`,
        borderHover: `gray.300`,
        ring: `${color}.500`,
      })
  }
}

export const useButtonStyle = (getter: MaybeRefOrGetter<ButtonInnerProps>) => {
  const { convertAliasColor } = useTheme()
  const result = computed(() => {
    const props = toValue(getter)
    return {
      class: createClass({
        variant: props.variant,
        color: props.color as 'primary',
        size: props.size,
        block: props.block,
        pill: props.pill || props.circle,
        square: props.square || props.circle,
        disabled: props.disabled,
      }),
      style: createStyle(props.variant, convertAliasColor(props.color)),
    }
  })
  return result
}
