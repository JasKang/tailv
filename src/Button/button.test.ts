import { mount } from '@vue/test-utils'
import { describe, expect, test } from 'vitest'
import Button from './index.vue'

describe('Button', () => {
  test('default', async () => {
    const wrapper = mount(Button, {
      props: {},
      slots: { default: 'Hello world' },
    })
    const el = wrapper.element as HTMLButtonElement

    // Assert the rendered text of the component
    expect(wrapper.text()).toContain('Hello world')

    expect(el.className.split(' ').includes('bg-input-background')).toBe(true)
  })

  test('variant', async () => {
    const wrapper = mount(Button, {
      props: { variant: 'outline' },
      slots: { default: 'Hello world' },
    })
    const el = wrapper.element as HTMLButtonElement
    expect(el.className.split(' ').includes('border-primary-border')).toBe(true)
  })
})
