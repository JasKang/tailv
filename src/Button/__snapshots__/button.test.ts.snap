// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > default 1`] = `
"<button type="button" class="focus-visible:outline-primary-solid inline-flex cursor-pointer appearance-none items-center justify-center whitespace-nowrap text-center font-medium outline-none transition-[color,background-color,text-decoration-color] focus:z-10 focus-visible:outline-2 focus-visible:outline-offset-2 disabled:cursor-not-allowed disabled:opacity-50 border-input-border text-foreground bg-input-background enabled:hover:bg-input-background-hover border shadow-sm h-9 text-sm py-1.5 px-3 rounded-md">
  <!--v-if-->Hello world
</button>"
`;
